### ISSUE TITLE
False positive for `__eq__` method on a generic

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#4232]
```
# pyright: strict

from typing import Generic, TypeVar

T = TypeVar("T")

class Foo(Generic[T]):

    value: T

    def __init__(self, value: T):
        self.value = value

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Foo):
            return False
        return self.value == other.value
```

The above yields two errors:

```
foo.py:18:30 - error: Type of "value" is unknown (reportUnknownMemberType)
foo.py:18:16 - error: Return type is unknown (reportUnknownVariableType)
```

**Expected behavior**
No errors (especially not `Return type is unknown`-- return type is definietly `bool`.

**VS Code extension or command-line**
Command line, version `1.1.280`

**Comment by erictraut** [comment]
The behavior is intended. The `isinstance` type narrowing logic narrows the type of `object` to `Foo[Unknown]` since the type argument cannot be determined. That means the type of `other.value` is `Unknown`. Any binary expression where one of the operands is type `Unknown` or `Any` propagates that type as the result.

If you want the type checker to assume that the type of `other` is `Foo[T]` rather than `Foo[Unknown]`, you could add an explicit cast.

```python
        o = cast(Foo[T], other)
        return self.value == o.value
```
